Blog post about optimization (#1780)
vitessio/website: content/en/blog/2024-07-02-an-interesting-optimisation.md

---
author: 'Andrés Taylor'
date: 2024-07-02

> slug: '2024-07-02-an-interesting-optimisation'

Reviewer: FYI, if we don't merge this today we should keep an eye out on those dates + the name of the file.

> slug: '2024-07-02-an-interesting-optimisation'

Reviewer 2: ```suggestion
slug: '2024-07-02-an-interesting-optimization'
```

tags: ['Vitess', 'PlanetScale', 'MySQL', 'Query Serving', 'Vindex', 'plan', 'execution plan', 'explain', 'optimizer']
title: 'An Interesting Optimization'

> description: "How I implemented an optimisation by delaying another optimisation"

Reviewer 2: ```suggestion
description: "How I implemented an optimization by delaying another optimization"
```

Reviewer 2: This word has been spelt both ways in the blog. The exact spelling doesn't matter as long as we keep it consistent.

---

## Introduction

I recently encountered an intriguing bug. A user reported that their query was causing vtgate to fetch a large amount of data, sometimes resulting in an Out Of Memory (OOM) error.

> For a deeper understanding of grouping and aggregations on Vitess, I recommend reading [this blog post](https://planetscale.com/blog/grouping-and-aggregations-on-vitess) I wrote earlier.

Reviewer 2: ```suggestion
For a deeper understanding of grouping and aggregations on Vitess, I recommend reading this [prior blog post](https://planetscale.com/blog/grouping-and-aggregations-on-vitess).
```

## Planning and Tree Rewriting

During planning, we perform extensive tree rewriting to push as much work down under `Route`s as possible. The rewriting process aims to optimize query execution by pushing operations closer to the data.

### Initial Plan

The first plan after horizon expansion looked like this:

```
Ordering (user_extra.id)
└── Aggregator (ORG sum(`user`.type), user_extra.id group by user_extra.id)
    └── ApplyJoin on [`user`.team_id | :user_team_id = user_extra.id | `user`.team_id = user_extra.id]
        ├── Route (Scatter on user)
        │   └── Table (user.user)
        └── Route (Scatter on user)
            └── Filter (:user_team_id = user_extra.id)
                └── Table (user.user_extra)
```

## Trying to Optimize the Plan

We don't split aggregation between MySQL and vtgate in the initial phases, so we couldn't immediately push down the aggregation through the join. However, we can push down ordering under the aggregation.

### Pushing Ordering Under Aggregation

By pushing ordering under aggregation, the plan changes to:

```

> >>>>>>>> push ordering under aggregation

Reviewer: do we want to keep this `>>>>` line since we don't show it in the other trees.

Aggregator (ORG sum(`user`.type), user_extra.id group by user_extra.id)
└── Ordering (user_extra.id)

>     └── ApplyJoin on [`user`.team_id | :user_team_id = user_extra.id | `user`.team_id = user_extra.id]

Reviewer: Might be worth expanding and showing everything under the `ApplyJoin` too, that way all the trees are consistent.

Reviewer 2: Or just `...` to show that the rest may not be interesting.

```

> We can't push the ordering further down since it's sorted by the RHS of the join. Ordering can only be pushed down to the LHS.

Reviewer: ```suggestion
We can't push the ordering further down since it's sorted by the right-hand-side of the join. Ordering can only be pushed down to the left-hand-side.
```

Reviewer 2: I like this suggestion, but remove the hyphens aka dashes. So `right hand side` and `left hand side`. Optionally (RHS) and (LHS), though not necessary because they aren't used ever again in the text of the blog.

This leaves us in an unfortunate situation - ordering is blocking the aggregator from being pushed down, which means we have to fetch all that data, _and_ sort it to do the aggregation.

### The Solution

The solution I typically use in these situations involves leveraging the phases we have in the planner.

### Phases

> We have several phases that run sequentially. After completing a phase, we run the push-down rewriters until we reach a fixed point, then move to the next phase, and so on.

Reviewer 2: What is the "fixed point"?

Rewriters perform one of two functions:

> 1. Running a rewriter over the plan to perform a specific task. For example, the "pull distinct from UNION" rewriter extracts the DISTINCT part from UNION and uses a separate operator for it.

Reviewer 2: ```suggestion
1. Running a rewriter over the plan to perform a specific task. For example, the "pull DISTINCT from UNION" rewriter extracts the DISTINCT part from UNION and uses a separate operator for it.
```

> 2. Enabling and disabling rewriters. Many rewriters are only enabled after reaching a certain point.

Reviewer: It's not clear to me if the rewriter we are currently running enables/disables itself or not.

Reviewer 2: Yeah, this is vague. Do rewriters enable and disable other rewriters?

> By delaying the "ordering under aggregation" rewriter until the "split aggregation" phase, we can push down the aggregation without causing issues by pushing down ordering under the top aggregator.

Reviewer: Ah now it makes sense why you would want to keep the `>>>>>` in the previous tree. Maybe instead of leaving the `>>>>` we can name the tree directly in the blog post.

Reviewer 2: What is the meaning of this? Is it:
1. "We can push down the aggregation" "without causing issues" by "pushing down ordering ...."
OR
2. "We can push down the aggregation" "without causing issues by pushing down ordering" by "delaying the ... rewriter"?
IOW, is it the delaying that solves the problem, or is it the "pushing down ordering ..."?
I understand that it is (2), but the sentence should be rewritten to make it crystal clear.
